Adding Basic auth and Request Header auth to Admin Solutions guide

tpoitras opened the pull request:
https://bugzilla.redhat.com/show_bug.cgi?id=1339838

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,308 @@ toc::[]
+[[BasicAuthPasswordIdentityProvider]]
+
+Configuring Basic Authentication allows users to log in to {product-title} with
+credentials stored on a remote identity provider.
```
This is not specific enough... all the auth integrations do that.

BasicAuthPasswordIdentityProvider and RequestHeaderIdentityProvider are the most generic integration mechanisms.

BasicAuthPasswordIdentityProvider is a backend integration. It makes users send their username/password to OpenShift, and OpenShift then verifies those credentials against an external server by making a server-to-server request, passing the given credentials as a basic auth header.
* This only works for username/password login mechanisms
* OpenShift must be able to make network requests to the remote auth server
* Requires users to send their credentials to OpenShift during login

RequestHeaderIdentityProvider makes users talk to OpenShift through a proxy, which is responsible for 1) authenticating the user any way it wants to, and 2) providing the user's identity to OpenShift via a request header.
* The proxy must be able to make network requests to the OpenShift server
* Unauthenticated login attempts are redirected to a configured proxy URL
* The proxy can authenticate browser clients any way it wants to, but must currently use basic auth or kerberos if it wants to work with the `oc` CLI tooling.
* Means user credentials can be handled by the proxy and never seen by OpenShift

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,308 @@ toc::[]
+
+[WARNING]
+====
+*[NEEDINFO]*: Is this troubleshooting information? It was in introduction of the install guide authentication topic, but seems out of place there. What purpose does this serve? Does any of it belong in the overview here, in other words does any of this need to be considered before configuring authentication?
```
Most troubleshooting will probably be around network connectivity to the back end server.

Could include equivalent `curl` commands that could be run on the master for debugging. Something like this would be the equivalent of what the OpenShift API server would do, given the sample config. Replace "sampleuser:samplepassword" with a valid username/password to test a valid login, and replace with an invalid username/password to test an invalid login.

```
curl --cacert /path/to/ca.crt --cert /path/to/client.crt --key /path/to/client.key -u sampleuser:samplepassword -v https://www.example.com/remote-idp
```

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,308 @@ toc::[]
+
+To add a user to {product-title}, the user must exist on the Basic
+authentication server, and if required you must create a new Basic
+authentication account for the user.
```
a "basic authentication account" is not a thing... "Basic authentication" refers to the way the credentials are sent from OpenShift to the remote auth server, nothing more

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,308 @@ toc::[]
+For users to authenticate using this identity provider, they must access
+_<master>/oauth/authorize_ via an authenticating proxy. You can either proxy the
+entire master API server so that all access goes through the proxy, or you can
+configure the OAuth server to redirect unauthenticated requests to the proxy.
```
I would not recommend proxying the entire API

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,308 @@ toc::[]
+
+[WARNING]
+====
+*NEEDINFO*: The value <master>/oauth/authorize, is that a URL and not a file path?
```
yes, a URL

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,308 @@ toc::[]
+.. Set the `*provider.loginURL*` parameter to the proxy URL to which to send clients expecting login flows. This parameter can include the 
+xref:../install_config/configuring_authentication.adoc#RequestHeaderIDP-urlquerytokens[`${url}` and `${query}` tokens]
+in the query portion of the URL. 
+.. Optionally, change the `*clientCA*` parameter to check incoming requests for a valid client certificate before the request's headers are checked for a user name. 
```
and `*clientCommonNames*`

liggitt replied:
Need to describe setting the headers that will be used for the identity, preferred username, display name, and email

liggitt replied:
I don't know that I would have them copy/paste that config as a starting point... I'd want them to decide what headers they were going to pass in, not start with a list designed to illustrate looking at multiple headers, when they might not want that

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,308 @@ toc::[]
++
+[WARNING]
+====
+If you expect unauthenticated requests to reach the OAuth server, a `*clientCA*`
```
and optionally `clientCommonNames`

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,308 @@ toc::[]
+----
+
+Once configured, any user logging in to the {product-title} web console will be
+prompted to log in using Request Header credentials.
```
"Request Header credentials" aren't a thing. They'll be redirected to the authenticating proxy, which will authenticate the user in some unspecified way.

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,308 @@ toc::[]
+defined there, and any user with a valid user name for the configured
+authentication server can log in.
+
+To add a user to {product-title}, the user must exist on the Request Header authentication server, and if required you must create a new Request Header authentication account for the user.
```
This has to be more vague... they'll need to add the users to whatever system the proxy is using to authenticate them.

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,308 @@ toc::[]
+
+[WARNING]
+====
+*[NEEDINFO]*: Do you have access to a request header authentication setup? If so, can you please provide me with the output for the `oc get users` command when request header is configured as the auth provider for OpenShift? Thank you.
```
the users that result from all the different auth methods are identical. they only vary based on the name that was chosen for the provider, and whether the provider makes display name info available.

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,308 @@ toc::[]
+
+[WARNING]
+====
+*[NEEDINFO]*: Do you have access to a basic authentication setup? If so, can you please provide me with the output for the `oc get users` command when basic auth is configured as the auth provider for OpenShift? Thank you.
```
this step doesn't make sense to repeat for every identity provider... `oc get users` looks the same, no matter what IDP you use.

tpoitras commented:
Made updates based on technical review from @liggitt

tpoitras commented:
Made minor update based on review from Takayoshi Kimura and Eido Inoue

tpoitras commented:
I'll squash before merging.

Ping @bfallonf @ahardin-rh @tnguyen-rh @adellape for Peer Review, please. :alien:

tpoitras commented:
Moved BZ along to QA, once it passes will do peer review.

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+.. If required,
+xref:../install_config/configuring_authentication.adoc#mapping-identities-to-users[change `*mappingMethod*`] to control how mappings are established between the
+provider's identities and user objects.
+.. Specify the `*url*` to use to connect to a server that accepts credentials in Basic authentication headers.
```
note here that this should be an HTTPS URL

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+====
+Basic Authentication is not a secure exchange in principle; it needs to be
+combined with encryption or certificate authentication to prevent potential
+snooping of the user ID and password, and to prevent man-in-the-middle attacks.
```
I understand what this is saying, but I would not lead with "Basic Authentication is not a secure exchange in principle". I would rephrase as "Basic Authentication should use a HTTPS connection to the remote server in order to prevent potential snooping of the user ID and password, and to prevent man-in-the-middle attacks."

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+[[BasicAuthPasswordIdentityProvider]]
+
+Basic Authentication is a generic backend integration mechanism that allows
+users to log in to {product-title} with credentials stored on a remote identity
```
s/stored on/validated against/

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+xref:../install_config/configuring_authentication.adoc#mapping-identities-to-users[change `*mappingMethod*`] to control how mappings are established between the
+provider's identities and user objects.
+.. Specify the `*url*` to use to connect to a server that accepts credentials in Basic authentication headers.
+.. Optionally, change the `*ca*` to the certificate bundle to use in order to validate server certificates for the configured URL.
```
note, "or leave it empty to use the system trusted roots"

liggitt replied:
none of the example sample paths are likely to work for them... I would say "set" instead of "change"

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+provider's identities and user objects.
+.. Specify the `*url*` to use to connect to a server that accepts credentials in Basic authentication headers.
+.. Optionally, change the `*ca*` to the certificate bundle to use in order to validate server certificates for the configured URL.
+.. Optionally, change the `*certFile*` to the client certificate to present when making requests to the configured URL.
```
set or remove

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+
+The proxy must be able to make network requests to the {product-title} server.
+Unauthenticated login attempts are redirected to a configured proxy URL. The
+proxy can authenticate browser clients however it is configured, but it must
```
"however it is configured" reads oddly.

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+to control how mappings are established between the provider's identities and
+user objects.
+.. Set the `*challenge*` parameter to *true* to redirect unauthenticated
+requests from clients expecting `WWW-Authenticate` challenges. 
```
"(like the CLI)"

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+.. Set the `*challenge*` parameter to *true* to redirect unauthenticated
+requests from clients expecting `WWW-Authenticate` challenges. 
+.. Set the `*provider.challengeURL*` parameter to the proxy URL to which to send
+clients expecting `WWW-Authenticate` challenges. 
```
", like the `oc` CLI client"

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+xref:../install_config/configuring_authentication.adoc#RequestHeaderIDP-urlquerytokens[`${url}` and `${query}` tokens]
+in the query portion of the URL. 
+.. Set the `*login*` parameter to *true* to redirect unauthenticated requests
+from clients expecting login flows. 
```
", like web browser clients"

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+.. Set the `*login*` parameter to *true* to redirect unauthenticated requests
+from clients expecting login flows. 
+.. Set the `*provider.loginURL*` parameter to the proxy URL to which to send
+clients expecting login flows. This parameter can include the 
```
", like web browser clients"

liggitt commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+clients expecting login flows. This parameter can include the 
+xref:../install_config/configuring_authentication.adoc#RequestHeaderIDP-urlquerytokens[`${url}` and `${query}` tokens]
+in the query portion of the URL. 
+.. Optionally, change the `*clientCA*` parameter to check incoming requests for
```
This should absolutely be done. Without checking the request containing user info in headers against a client CA, anyone that can hit the API server can assume any identity they want

liggitt replied:
I would remove the "Optionally", and just say "Change the clientCA parameter to the certificate bundle to use..."

tpoitras commented:
Thanks for the feedback, @liggitt - I have incorporated it.

tpoitras commented:
Peer Review, please: @bfallonf @ahardin-rh @adellape

tpoitras commented:
bump @bfallonf @ahardin-rh @adellape

bfallonf commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+=== Configuring Authentication on the Master
+
+. If you have:
+.. Already completed the installation of Openshift, then copy the
```
These substeps (1a and 1b) are options rather than substeps, correct? Can they be marked up with points rather than steps? It's one or the other, so not sure if numbered list is right?

bfallonf commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+. Edit the new *_master-config.yaml_* file's `*identityProviders*` stanza.
+. Copy
+xref:../install_config/configuring_authentication.adoc#basic-auth-example-config[the example `*BasicAuthPasswordIdentityProvider*` configuration] and paste it to replace the existing stanza.
+. Make the following modifications to the `*identityProviders*` stanza:
```
Same comment as above

tpoitras replied:
no, the comment you made above was valid, but these are required steps aside from the "optionally" ones, and we still number "optional" steps in our enumerated lists, so, I'm leaving it as is.

bfallonf commented on admin_solutions/authentication.adoc:
```diff
@@ -5,10 +5,303 @@ toc::[]
+=== Configuring Authentication on the Master
+
+. If you have:
+.. Already completed the installation of Openshift, then copy the
```
Gah, same comment!

bfallonf commented:
@tpoitras Largely great. A few comments about list-type stuff. Otherwise, LGTM :peach:

tpoitras commented:
Thanks for the review, @bfallonf , and good catch on that list formatting.

I've incorporated your feedback, and pushed the changes (now squashed flat to one commit)

Rendered docs updated: http://file.bne.redhat.com/tpoitras/2016/basicrequest/openshift-enterprise/basicrequest-BZ1339838/admin_solutions/authentication.html

Merging for Next Release.

tpoitras commented:
[rev_history]
|link:../admin_solutions/authentication.html[Source-to-Image (S2I) -> Ruby]
|Added xref:../admin_solutions/authentication.adoc#basic-auth-remote[Basic Auth] and xref:../admin_solutions/authentication.adoc#request-header-auth[Request Header] to the Authentication topic.
%
